### PR TITLE
[NTDLL] Improve exports

### PR DESCRIPTION
**Pull request by tkreuzer**
* [NTDLL] Ordinals -> @
* [NTDLL] Export some more existing functions
* [RTL][NTDLL] Add some function stubs
* [NTDLL] Add some more stub exports
* [NTDLL] Remove redundant Zw* -> Nt* forwarders

### CODE REVIEW

**Comment by HBelusca** [comment]
Regarding "Ordinals -> @" it was done in the past in order to be sure we match 100% Windows Server 2003.

File: dll/ntdll/def/ntdll.spec [hunk 1]
```diff
@@ -7,9 +7,9 @@
 @ fastcall -arch=i386 RtlUlongByteSwap(long)
 @ fastcall -ret64 RtlUlonglongByteSwap(double)
 @ fastcall -arch=i386 RtlUshortByteSwap(long)
-@ stdcall -arch=i386,x86_64 ExpInterlockedPopEntrySListEnd()
-@ stdcall -arch=i386,x86_64 ExpInterlockedPopEntrySListFault()
-@ stdcall -arch=i386,x86_64 ExpInterlockedPopEntrySListResume()
+@ stdcall -arch=i386 ExpInterlockedPopEntrySListEnd()
```
Comment (HBelusca) [inline dll/ntdll/def/ntdll.spec]: Why do we need them splitted?

File: dll/ntdll/ldr/ldrapi.c [hunk 2]
```diff
@@ -1658,4 +1658,19 @@ LdrFlushAlternateResourceModules(VOID)
     return FALSE;
 }
 
+/*
+ * @unimplemented
+ * See https://www.kernelmode.info/forum/viewtopic.php?t=991
+ */
+NTSTATUS
+NTAPI
+LdrSetAppCompatDllRedirectionCallback(
```
Comment (learn-more) [inline dll/ntdll/ldr/ldrapi.c]: I never looked at this because in vista+ it's gone.
Do you have plans with this?

**Comment by tkreuzer** [comment]
Now that all functions are exported, it should not make any difference. Also it doesn't make sense to use ordinals, because it doesn't match between different architecture, different versions, not even between different SPs, making import by ordinal absolutely impossible for any application. So whether you use these ordinals or some others doesn't make any difference. I do not plan to maintain one spec file for every architecture times every major version, because that would be unmaintainable. This is just another case of "If it ain't broke, don't fix it til it is." I don't see anything to fix here.

File: dll/ntdll/ldr/ldrapi.c (hunk 2 above)
Comment (tkreuzer) [reply dll/ntdll/ldr/ldrapi.c]: No, I just googled some functions for their prototypes and since I found this, I added a stub.

**Comment by learn-more** [comment]
Only one where we should pay attn (that I know of) is ws2_32 since a lot of shit imports by ordinal from that.

File: dll/ntdll/def/ntdll.spec (hunk 1 above)
Comment (tkreuzer) [reply dll/ntdll/def/ntdll.spec]: It was an attempt to get the order right, but in fact dlltool reorders them in the "wrong way". MS orders exports by symbol name, GCC/dlltool orders them somewhat by export name (in fact the logic is screwed, as it first does a qsort by export name, then a qsort that only sorts ordinal exports vs auto-ordinal). Proper solution would be to fix dlltool, workaround could be spec2def forcing ordinals, I'm looking into it...

**Comment by pithwz** [comment]
hi,
If it's possible please add these API missing from Windows 2003 SP2  v.5.2.3790

LdrAlternateResourcesEnabled
LdrFindResourceEx_U
LdrHotPatchRoutine
LdrInitShimEngineDynamic
LdrQueryImageFileExecutionOptionsEx
LdrSetAppCompatDllRedirectionCallback
PfxFindPrefix
PfxInitialize
PfxInsertPrefix
PfxRemovePrefix
RtlAddCompoundAce
RtlAddressInSectionTable
RtlAppendPathElement
RtlCancelTimer
RtlCaptureStackContext
RtlCheckProcessParameters
RtlDebugPrintTimes
RtlEnableEarlyCriticalSectionEventCreation
RtlGetUnloadEventTrace
RtlInitializeAtomPackage
RtlInsertElementGenericTableFull
RtlInsertElementGenericTableFullAvl
RtlLogStackBackTrace
RtlLookupElementGenericTableFull
RtlLookupElementGenericTableFullAvl
RtlMultipleAllocateHeap
RtlMultipleFreeHeap
RtlQueryProcessBackTraceInformatio
RtlQueryProcessHeapInformation
RtlQueryProcessLockInformation
RtlQueueApcWow64Thread
RtlSetTimer
RtlSetUnicodeCallouts
RtlTraceDatabaseAdd
RtlTraceDatabaseCreate
RtlTraceDatabaseDestroy
RtlTraceDatabaseEnumerate
RtlTraceDatabaseFind
RtlTraceDatabaseLock
RtlTraceDatabaseUnlock
RtlTraceDatabaseValidate
RtlUnhandledExceptionFilter2
RtlpNotOwnerCriticalSection

**Comment by tkreuzer** [comment]
> 
> 
> hi,
> If it's possible please add these API missing from Windows 2003 SP2 v.5.2.3790

What are you talking about? All of them are in the spec file. Some of them are stubs, but that has nothing to do with this commit.

**Pull request merged**